dbms/src/Storages/DeltaMerge/File/DMFileWithVectorIndexBlockInputStream.cpp

```diff
@@ -0,0 +1,487 @@
+
+    // Skip as many packs as possible according to Pack Filter
+    while (index_reader_next_pack_id < all_packs)
+    {
+        if (isUse(pack_res[index_reader_next_pack_id]))
+            break;
+        index_reader_next_row_id += pack_stats[index_reader_next_pack_id].rows;
+        index_reader_next_pack_id++;
+    }
+
+    if (index_reader_next_pack_id >= all_packs)
+        // Finished
+        return {};
+
+    auto read_pack_id = index_reader_next_pack_id;
+    auto block_start_row_id = index_reader_next_row_id;
+    auto read_rows = pack_stats[read_pack_id].rows;
+
+    index_reader_next_pack_id++;
+    index_reader_next_row_id += read_rows;
+
+    Block block;
+    block.setStartOffset(block_start_row_id);
+
+    auto vec_column = vec_cd.type->createColumn();
+    vec_column->reserve(read_rows);
+
+    Stopwatch w;
+    vec_column_reader->read(vec_column, read_pack_id, read_rows);
+    duration_read_from_vec_index_seconds += w.elapsedSeconds();
+
+    block.insert(ColumnWithTypeAndName{//
+                                       std::move(vec_column),
+                                       vec_cd.type,
+                                       vec_cd.name,
+                                       vec_cd.id});
+
+    return block;
+}
+
+Block DMFileWithVectorIndexBlockInputStream::readByFollowingOtherColumns()
+{
+    // First read other columns.
+    Stopwatch w;
+    auto block_others = reader.read();
+    duration_read_from_other_columns_seconds += w.elapsedSeconds();
+
+    if (!block_others)
+        return {};
+
+    auto read_rows = block_others.rows();
+
+    // Using vec_cd.type to construct a Column directly instead of using
+    // the type from dmfile, so that we don't need extra transforms
+    // (e.g. wrap with a Nullable). vec_column_reader is compatible with
+    // both Nullable and NotNullable.
+    auto vec_column = vec_cd.type->createColumn();
+    vec_column->reserve(read_rows);
+
+    // Then read from vector index for the same pack.
+    w.restart();
+
+    vec_column_reader->read(vec_column, getPackIdFromBlock(block_others), read_rows);
+    duration_read_from_vec_index_seconds += w.elapsedSeconds();
+
+    // Re-assemble block using the same layout as header_layout.
+    Block res = header_layout.cloneEmpty();
+    // Note: the start offset counts from the beginning of THIS dmfile. It
+    // is not a global offset.
+    res.setStartOffset(block_others.startOffset());
+    for (const auto & elem : block_others)
+    {
+        RUNTIME_CHECK(res.has(elem.name));
+        res.getByName(elem.name).column = std::move(elem.column);
+    }
+    RUNTIME_CHECK(res.has(vec_cd.name));
+    res.getByName(vec_cd.name).column = std::move(vec_column);
+
+    return res;
+}
+
+void DMFileWithVectorIndexBlockInputStream::load()
+{
+    if (loaded)
+        return;
+
+    Stopwatch w;
+
+    loadVectorIndex();
+    loadVectorSearchResult();
+
+    duration_load_vec_index_and_results_seconds = w.elapsedSeconds();
```

Comment:
What's the usage of this metric? It adds up downloading(maybe) the vector index from S3 with building the result set. There is too much work here?

Reply:
The overall duration of read